### ISSUE TITLE
incorrect calculation of mAP for -points 0 (for  own dataset)

### ISSUE THREAD

**Issue by adityap27** [AlexeyAB/darknet#6040]
@AlexeyAB 

   * The code for calculation of Area-Under-Curve on PR-chart.(for -point 0) seems wrong.

   * command:
**./darknet detector map "data/obj.data" "cfg/yolov3.cfg" "backup/yolov3_last.weights" -dont_show -thresh 0.01 -points 0**

Screenshot of code in **detector.c**
![image](https://user-images.githubusercontent.com/30194525/85260699-bb59d500-b488-11ea-9f0a-c8bed77f468d.png)

Problem: the code is not calculating the area of section between recall=0 and 1st recall value(from left on graph). i.e delta recall is only calculated uptill top 2 ranks.......not between the 1st rank and recall=0. 

This problem occurs only if the recall at rank 1 is non-zero........in that case we need to calculate another delta_recall which isn't calculated.

output
![image](https://user-images.githubusercontent.com/30194525/85261414-c06b5400-b489-11ea-8e2c-6902ffee100d.png)

**Comment by adityap27** [comment]
Hey @AlexeyAB  I have solved this incorrect calculation by changing the detector.c file.

**May I create a pull request** so that map is calculated accurately ?? (I have noticed that old map calculation wasn't correct for less images so...I solved it for less images too.)

In next comment I have added **comparison** of mAP for old_detector.c vs new detector.c

**Comment by adityap27** [comment]
### OLD CODE
![image](https://user-images.githubusercontent.com/30194525/86606789-167de280-bfc6-11ea-9a70-d626d331c26c.png)

### NEW CODE
![image](https://user-images.githubusercontent.com/30194525/86606890-34e3de00-bfc6-11ea-9e16-6588ec3d1fd4.png)

**My Actual image that is used for mAP calculation.**
![image](https://user-images.githubusercontent.com/30194525/86606942-462cea80-bfc6-11ea-9d07-e79d1b99c4ec.png)


We can clearly see that the aP for class mask should be 100% aP .....but the old code was giving 75% aP due to the reason I stated above.(less images)

**Comment by AlexeyAB** [comment]
* Do you mean that you get Area under PR-curve = 1 for all confidence thresh  >= 0.005 ? https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L951

* Add between these lines: https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L1255-L1256
this line 
`printf("class_id = %d, point = %d, cur_recall = %.4f, cur_precision = %.4f \n", i, point, last_recall, last_precision);`
and show me screenshot

* Yes, create a Pull request with minimal changes and refer to this issue.

**Comment by adityap27** [comment]
For 1st question,

Yes, according to my test image, I should get area under PR-curve=1 for all conf thresh >=0.005, but due the old detector.c code it gave me incorrect aP(0.75) and mAP.

**Comment by adityap27** [comment]
> * Add between these lines: https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L1255-L1256
>   
>   this line
>   `printf("class_id = %d, point = %d, cur_recall = %.4f, cur_precision = %.4f \n", i, point, last_recall, last_precision);`
>   and show me screenshot

hey @AlexeyAB ,

Adding this line gives me error due to point variable is not accessible in that scope....as it is not there for map calculation for custom dataset.

![image](https://user-images.githubusercontent.com/30194525/86620006-38349500-bfd9-11ea-82e9-c68ed8ad28f8.png)

**Comment by AlexeyAB** [comment]
Try this line
`printf("class_id = %d, rank = %d, cur_recall = %.4f, cur_precision = %.4f \n", i, rank, last_recall, last_precision);`

**Comment by adityap27** [comment]
> printf("class_id = %d, rank = %d, cur_recall = %.4f, cur_precision = %.4f \n", i, rank, last_recall, last_precision);

Hey @AlexeyAB , Tried this new line....checkout the **screenshot below**

![image](https://user-images.githubusercontent.com/30194525/86621521-f6f1b480-bfdb-11ea-8ef2-d57fdcc0c11d.png)

**Comment by AlexeyAB** [comment]
Try to add this line between these two lines: https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L1254-L1255

**Comment by adityap27** [comment]
> Try to add this line between these two lines:
> 
> https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L1254-L1255

Hey @AlexeyAB , **check it out**
![image](https://user-images.githubusercontent.com/30194525/86623262-e55ddc00-bfde-11ea-9f28-0b38839d1053.png)

**note: this output is using my new detector.c ..........The old one would have given me 75%aP for class_id=1**

**Comment by AlexeyAB** [comment]
So we have:
r = 0.75  p = 1.0
r = 0.50  p = 1.0
r = 0.25  p = 1.0

So `area_under_curve = 0.75` so should be AP=75%

**Comment by AlexeyAB** [comment]
Thanks!
Also add this line between these 2 lines: https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L1244-L1245

Do you get cur_precision=1.0 for rank=3?

**Comment by adityap27** [comment]
> Thanks!
> Also add this line between these 2 lines:
> 
> https://github.com/AlexeyAB/darknet/blob/00919eeb99f73484cf9f51b277a4e4ebacef70df/src/detector.c#L1244-L1245
> 
> Do you get cur_precision=1.0 for rank=3?

@AlexeyAB  see now:

![image](https://user-images.githubusercontent.com/30194525/86625129-31f6e680-bfe2-11ea-94da-cc99c297dbff.png)

**Comment by AlexeyAB** [comment]
@adityap27 Thanks!

**Comment by adityap27** [comment]
> @adityap27 Thanks!

**Thank you @AlexeyAB  for your quick replies. :)**

**Comment by Fetulhak** [comment]
> > @adityap27 Thanks!
> 
> **Thank you @AlexeyAB for your support. :)**

Hi @adityap27 
can I use test set data to evaluate the performance of YOLOV3 model using mAP?

The other question I have is on the original repo in the file OBJ.data it says "valid" = test.txt 
does this mean the model is using my test data for model validation. I think in the machine learning course we learn models should be evaluated on validation set data.

**Comment by adityap27** [comment]
> > > @adityap27 Thanks!
> > 
> > 
> > **Thank you @AlexeyAB for your support. :)**
> 
> Hi @adityap27
> can I use test set data to evaluate the performance of YOLOV3 model using mAP?
> 
> The other question I have is on the original repo in the file OBJ.data it says "valid" = test.txt
> does this mean the model is using my test data for model validation. I think in the machine learning course we learn models should be evaluated on validation set data.

Hi @Fetulhak ,
1.)
Yes you can evaluate on test set after you have a weights file.
Code for evaluation on test set. (give .txt file of test set at "valid=" in .obj, before running command.)
```
#Performance on test set
./darknet detector map "darknet/obj.data" "darknet/cfg/yolov3.cfg" "darknet/backup/yolov3_final.weights" -dont_show -thresh 0.45
```

Although darknet will evaluate on 2 sets while training: train-set and valid-set.(specified by train= & valid=)
classes = 2
train  = data/train.txt
valid  = data/test.txt (or data/valid.txt)

Incase you have 3rd set (Test set), you can modify .obj file set valid=test.txt and evaluate using final weights at the end.


2.)
There are few confusions regarding 2 or 3 sets.

- -> Ideally we have 3 sets. Train, Valid, Test.
- -> Also , when using 2 sets, most people consider Validation Set = Test Set and use both names interchangeable.(as seen here in darknet)

Here in darknet (and many other people) use only 2 sets which is completely fine when your 2nd set (valid) is pretty large and thus you don't need test set.

But incase you have 3 sets(I also used 3 sets in my case), then you can follow the command I gave above. You can use 2 sets for evaluation on the go while training and after it gets completed modify .obj for valid=test-set.txt and run command for evaluation or test set :) .

Hope this helps you.

**Comment by Fetulhak** [comment]
> > > @adityap27 Thanks!
> > 
> > 
> > **Thank you @AlexeyAB for your support. :)**
> 
> Hi @adityap27
> can I use test set data to evaluate the performance of YOLOV3 model using mAP?
> 
> The other question I have is on the original repo in the file OBJ.data it says "valid" = test.txt
> does this mean the model is using my test data for model validation. I think in the machine learning course we learn models should be evaluated on validation set data.

Hi @adityap27

Thanks. You make it clear to me how the three datasets can be used. 

The other concern I have seen on YOLOV3 is that changing the value of Conf_thresh and IOU_thresh manually as you did it above by giving conf_thresh value of 0.45.  The confusion I am having is that when I change the IOU_thres the mAP changes but when I change the Conf_thresh only precision, recall, F1-score change. But the thing is the mAP is  AUC of P-R and if precision and recall changes how it is not going to be affected i.e mAP?

**Comment by adityap27** [comment]
> > > > @adityap27 Thanks!
> > > 
> > > 
> > > **Thank you @AlexeyAB for your support. :)**
> > 
> > 
> > Hi @adityap27
> > can I use test set data to evaluate the performance of YOLOV3 model using mAP?
> > The other question I have is on the original repo in the file OBJ.data it says "valid" = test.txt
> > does this mean the model is using my test data for model validation. I think in the machine learning course we learn models should be evaluated on validation set data.
> 
> Hi @adityap27
> 
> Thanks. You make it clear to me how the three datasets can be used.
> 
> The other concern I have seen on YOLOV3 is that changing the value of Conf_thresh and IOU_thresh manually as you did it above by giving conf_thresh value of 0.45. The confusion I am having is that when I change the IOU_thres the mAP changes but when I change the Conf_thresh only precision, recall, F1-score change. But the thing is the mAP is AUC of P-R and if precision and recall changes how it is not going to be affected i.e mAP?

Hi @Fetulhak ,

I can answer regarding your confusion. First of all sorry for late reply, as I was busy last week.

Coming to your confusion that, on changing **conf_thresh**, why only precision, recall and F1- score changes but why mAP doesn't change.

1. It is completely normal that mAP doesn't change. Reason: it **doesn't depend on your given conf_thresh** and it considers **ALL** the predictions made by model. Check the below source where pre-defined thresh is used for mAP, in-order to consider all predictions. (0.005 is kept to avoid predictions with very very small conf.)
https://github.com/AlexeyAB/darknet/blob/64efa721ede91cd8ccc18257f98eeba43b73a6af/src/detector.c#L995

2. Precision, re-call, F1-score show to you is based on given conf_thresh, and thus considers only few predictions.

So to summarize, mAP calculation considers **all predictions**, whereas precision, recall, F1-score considers only **few predictions** (only those above given conf_thresh.) and so mAP will remain same irrespective of your given conf_thresh during evaluation.

I hope I was able to clear your confusion. Let me know if I can help you further. :)
